Lib/tomllib/mypy.ini

```
@@ -0,0 +1,17 @@
+# Config file for running mypy on tomllib.
+# Run mypy by invoking `mypy --config-file Lib/tomllib/mypy.ini`
+# on the command-line from the repo root
+
+[mypy]
+files = Lib/tomllib
+mypy_path = $MYPY_CONFIG_FILE_DIR/../../Misc/mypy
+explicit_package_bases = True
+python_version = 3.12
+pretty = True
+
+# Enable most stricter settings
+enable_error_code = ignore-without-code
+strict = True
+strict_bytes = True
+local_partial_types = True
+warn_unreachable = True
```

It feels like this PR is doing two things at once:
1. Adding a mypy check to CI
2. Making the type annotations in the code more precise

I think these should be separated into two PRs. If you make this change to the config file, I believe mypy should pass on `tomllib` as-is. You could then make a followup PR that removes `disallow_any_generics = False` from the config file again and applies the changes you're currently making to the type annotations in `Lib/tomllib`.

```suggestion
warn_unreachable = True
disallow_any_generics = False
```

---

Ok, sound reasonable!

---

Ok, sound reasonable!

---

Done!